packages/user_packages/iotcc_mqtt_device.py

```diff
@@ -0,0 +1,190 @@
+                registry.register(device_registry_name, reg_device)
+
+                # Setting multiple properties by passing Dictonary object for Devices with the retry attempts
+                # in case of exceptions
+                prop_attempts = 0
+                while prop_attempts <= retry_attempts:
+                    try:
+                        self.iotcc.set_properties(reg_device,
+                                                  {"Country": "USA-G", "State": "California", "City": "Palo Alto",
+                                                   "Location": "VMware HQ", "Building": "Promontory H Lab",
+                                                   "Floor": "First Floor"})
+                        break
+                    except Exception as e:
+                        prop_attempts = prop_attempts + 1
+                        log.error('Exception while setting property for Device {0} - {1}'.format(
+                            (device.name, str(e))))
+                        log.info('Trying setting properties for Device {0}: Attempt - {1}'.format(device.name,
+                                                                                                  str(prop_attempts)))
+                        time.sleep(delay_retries)
+
+                try:
+                    # Registering Metric for Device
+                    self.metrics = []
+                    metric_name = "Simulated Metrics"
+                    metric_simulated_received = Metric(name=metric_name, unit=None, interval=300,
+                                                       aggregation_size=1, sampling_function=device_metric)
+                    reg_metric_simulated_received = self.iotcc.register(metric_simulated_received)
+                    self.iotcc.create_relationship(reg_device, reg_metric_simulated_received)
+                    reg_metric_simulated_received.start_collecting()
+                    self.metrics.append(reg_metric_simulated_received)
+                except Exception as e:
+                    log.error(
+                        'Exception while loading metric {0} for device {1} - {2}'.format(metric_name, device.name,
+                                                                                         str(e)))
+
+            except Exception:
+                log.info("Device Registration and Metrics loading failed")
+                raise
+
+    def clean_up(self):
+        """
+        The clean up function of a liota package.
+        Unregister Device and Stops metric collection
+        :return:
+        """
+        # On the unload of the package the device will get unregistered and the entire history will be deleted
+        # from Pulse IoT Control Center so comment the below logic if the unregsitration of the device is not required
+        # to be done on the package unload
+        for device in self.reg_devices:
+            self.iotcc.unregister(device)
+            for metric in self.metrics:
```

Comment:
Sorry for noticing this late, but should we not stop collecting metrics first and then unregister the device?

Reply:
Changed

Reply:
Now that you moved the unregister call later, i notice that all the metrics of all the devices are being collected in a single self.metrics, but while stopping the collection, it is being called for each device. Hence, both these loops should be separate. :-(